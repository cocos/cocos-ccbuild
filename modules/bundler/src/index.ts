import * as rollup from './rollup';

export {
    rollup,
};