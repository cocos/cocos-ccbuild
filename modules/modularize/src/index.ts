export * from './module-config';
export * from './module-query';
export * from './platform-config';