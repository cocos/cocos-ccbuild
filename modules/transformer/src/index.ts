import * as babel from './babel';
import * as typescript from './typescript';

export { 
    babel,
    typescript,
};
