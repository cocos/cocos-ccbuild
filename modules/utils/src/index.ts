export * from './path';

export function asserts(expr: unknown, message?: string): boolean {
    if (!expr) {
        throw new Error(message);
    }
    return true;
}

export function isThenable (value: any) {
    // https://stackoverflow.com/a/53955664/10602525
    return Boolean(value && typeof value.then === 'function');
}