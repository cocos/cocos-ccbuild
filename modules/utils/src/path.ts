import * as ps from 'path';
import * as fs from 'fs-extra';

export function dirname (path: string): string {
    return formatPath(ps.dirname(path));
}

export function join (...args: string[]): string {
    return formatPath(ps.join(...args));
}

export function relative (from: string, to: string): string {
    return formatPath(ps.relative(from, to));
}

export function formatPath (path: string): string {
    return path.replace(/\\/g, '/');
}

export function absolutePathFuncFactory (dirname: string): (relativePath: string) => string {
    return function absolutePath (relativePath: string) {
        return join(dirname, relativePath);
    };
}

export function replaceExtname (path: string, extname: string): string {
    return path.slice(0, -ps.extname(path).length) + extname;
}

export function rebasePath (path: string, originDir: string, rebaseDir: string): string {
    return join(rebaseDir, ps.relative(originDir, path));
}

export function filePathToModuleRequest (path: string): string {
    return path.replace(/\\/g, '\\\\');
}

export function toExtensionLess (path: string): string {
    return path.slice(0, -ps.extname(path).length);
}

export async function readdirR (item: string, reduceOutput: string[]): Promise<void> { 
    if ((await fs.stat(item)).isDirectory()) {
        const dirItems = await fs.readdir(item);
        for (const subItem of dirItems) {
            await readdirR(join(item, subItem), reduceOutput);
        }
    } else {
        reduceOutput.push(item);
    }
}
