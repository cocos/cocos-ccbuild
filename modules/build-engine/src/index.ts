import { StatsQuery } from '@ccbuild/stats-query';
import { buildJsEngine } from './engine-js';
import { babel } from '@ccbuild/transformer';
import fs from 'fs-extra';
import { buildTsEngine } from './engine-ts';

function verifyCache (options: buildEngine.Options): boolean {
	// TODO
	return false;
}

function applyDefaultOptions (options: buildEngine.Options): void {
    options.preserveType ??= false;
    options.loose = true;  // force using true
    if (!options.flags) {
        options.flags = {};
    }
    switch (options.nativeCodeBundleMode) {
        case 'asmjs':
            options.flags.NATIVE_CODE_BUNDLE_MODE = 0;
            break;
        case 'wasm':
            options.flags.NATIVE_CODE_BUNDLE_MODE = 1;
            break;
        default:
            options.flags.NATIVE_CODE_BUNDLE_MODE = 2;
            break;
    }
}

function moduleOptionsToBabelEnvModules(moduleOptions: buildEngine.ModuleFormat): false | 'commonjs' | 'amd' | 'umd' | 'systemjs' | 'auto' {
    switch (moduleOptions) {
    case 'cjs': return 'commonjs';
    case 'system': return 'systemjs';
    case 'iife':
    case 'esm': return false;
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    default: throw new Error(`Unknown module format ${moduleOptions}`);
    }
}

/**
 * @group Merged Types
 */
export async function buildEngine (options: buildEngine.Options): Promise<buildEngine.Result> {
	applyDefaultOptions(options);

	if (verifyCache(options)) {
		throw 'TODO';
	}

	if (options.platform === 'OPEN_HARMONY') {
		if (options.preserveType) {
			// we use a custom engine builder for OPEN_HARMONY platform when enable preserveType option.
			return buildTsEngine(options);
		} else {
			return buildJsEngine(options as Required<buildEngine.Options>);
		}
	} else {
		if (options.preserveType) {
			console.warn(`Currently we haven't support building ts engine on the platform ${options.platform}`);
		}
		return buildJsEngine(options as Required<buildEngine.Options>);
	}
}

/**
 * @group Merged Types
 */
export namespace buildEngine {
    export type ModuleFormat = 'esm' | 'cjs' | 'system' | 'iife';
    
    export interface Options {
        /**
         * 引擎仓库目录。
         */
        engine: string;

        /**
         * 输出目录。
         */
        out: string;

        // TODO: should we provide IModeConfig because sometimes engine executes in multiple modes such as EDITOR and PREVIEW mode.
        mode: StatsQuery.ConstantManager.ModeType;

        platform: StatsQuery.ConstantManager.PlatformType;

        flags?: Partial<StatsQuery.ConstantManager.IFlagConfig>;

        /**
         * 包含的功能。
         */
        features?: string[];

        /**
         * 输出模块格式。
         * @default 'system'
         */
        moduleFormat?: ModuleFormat;

        /**
         * 是否对生成的脚本进行压缩。
         * @default false
         */
        compress?: boolean;

        /**
         * 是否生成 source map。
         * 若为 `inline` 则生成内联的 source map。
         * @default false
         */
        sourceMap?: boolean | 'inline';

        /**
         * 若 `sourceMap` 为 `true`，此选项指定了 source map 的路径。
         * @default `${outputPath.map}`
         */
        sourceMapFile?: string;

        /**
         * 若为 `true`，分割出 **所有** 引擎子模块。
         * 否则，`.moduleEntries` 指定的所有子模块将被合并成一个单独的 `"cc"` 模块。
         * @default false
         */
        split?: boolean;

        /**
         * 原生代码的打包模式
         * - 为 `wasm` 时使用 wasm 版本原生库
         * - 为 `asmjs` 时使用 asmjs 版本的原生库
         * - 为 `both` 时同时在结果中包含 wasm 与 asmjs 两个版本的原生库
         */
        nativeCodeBundleMode?: 'wasm' | 'asmjs' | 'both';

        /**
         * Wasm compression mode, 'brotli' means to compress .wasm to .wasm.br.
         * @note Currently, only WeChat and ByteDance mini-game support to load '.wasm.br' file.
         */
        wasmCompressionMode?: 'brotli';

        /**
         * If true, all deprecated features/API are excluded.
         * You can also specify a version range(in semver range) to exclude deprecations in specified version(s).
         * @default false
         */
        noDeprecatedFeatures?: string | boolean;

        /**
         * Experimental.
         */
        incremental?: string;

        // progress?: boolean;

        /**
         * BrowsersList targets.
         */
        targets?: string | string[] | Record<string, string>;

        /**
         * Enable loose compilation.
         * 
         * @deprecated since 1.1.20, we force using true internal.
         */
        loose?: boolean;

        /**
         * How to generate the reference to external assets:
         * - `'relative-from-out'`
         * Generate the path relative from `out` directory, does not contain the leading './'.
         *
         * - `'relative-from-chunk'`
         * Generate the path relative from the referencing output chunk.
         *
         * - `'dynamic'`(default)
         * Use runtime `URL` API to resolve the absolute URL.
         * This requires `URL` and `import.meta.url` to be valid.
         */
        assetURLFormat?: 'relative-from-out' | 'relative-from-chunk' | 'runtime-resolved';

        /**
         * Preserve engine type info, this options will build a TS engine to the output directory.
         * It's useful when we need to take a step towards the AOT optimization.
         * This options is only supported on Open Harmony platform for now.
         * @default false
         */
        preserveType?: boolean;

        visualize?: boolean | {
            file?: string;
        };

        // buildTimeConstants: IBuildTimeConstants;

        /**
         * Generate cocos/native-binding/decorators.ts for native platforms
         */
        generateDecoratorsForJSB?: boolean;

        // /**
        //  * Whether force SUPPORT_JIT to the specified value.
        //  */
        // forceJitValue?: boolean,

        /**
         * Whether to generate 'named' register code for systemjs module format.
         * SystemJS default register code: System.register([], function(){...});
         * SystemJS named register code: System.register('module_name', [], function(){...});
         * @note It's only avaiable when options.moduleFormat is 'system'.
         */
        enableNamedRegisterForSystemJSModuleFormat?: boolean;
    }

    export interface Result {
        /**
         * Mappings between feature unit name and their actual chunk file, for example:
         * ```js
         * {
         *   "core": "./core.js",
         *   "gfx-webgl": "./gfx-webgl.js",
         * }
         * ```
         */
        exports: Record<string, string>;

        /**
         * The compulsory import mappings that should be applied.
         */
        chunkAliases: Record<string, string>;

        /**
         * The dependency graph, only including dependency chunks.
         * 
         * @deprecated please use `chunkDepGraph` instead.
         */
        dependencyGraph?: Record<string, string[]>;
        
        chunkDepGraph: Record<string, string[]>;

        assetDepGraph: Record<string, string[]>;

        hasCriticalWarns: boolean;
    }

    export async function transform(code: string, moduleOption: ModuleFormat, loose?: boolean): Promise<{ code: string; }> {
        const babelFormat = moduleOptionsToBabelEnvModules(moduleOption);
        const babelFileResult = await babel.core.transformAsync(code, {
            presets: [[babel.presets.presetEnv, { modules: babelFormat, loose: loose ?? true } as babel.presets.presetEnv.Options]],
        });
        if (!babelFileResult || !babelFileResult.code) {
            throw new Error('Failed to transform!');
        }
        return {
            code: babelFileResult.code,
        };
    }

    export async function isSourceChanged(incrementalFile: string): Promise<boolean> {
        let record: Record<string, number>;
        try {
            record = await fs.readJSON(incrementalFile);
        } catch {
            console.debug(`Failed to read incremental file: ${incrementalFile} - rebuild is needed.`);
            return true;
        }
        for (const file of Object.keys(record)) {
            const mtime = record[file];
            try {
                /* eslint-disable-next-line no-await-in-loop */
                const mtimeNow = (await fs.stat(file)).mtimeMs;
                if (mtimeNow !== mtime) {
                    console.debug(`Source ${file} in watch files record ${incrementalFile} has a different time stamp - rebuild is needed.`);
                    return true;
                }
            } catch {
                console.debug(`Failed to read source ${file} in watch files record ${incrementalFile} - rebuild is needed.`);
                return true;
            }
        }
        return false;
    }

    // eslint-disable-next-line no-inner-declarations
    function _enumerateDependentChunks (meta: buildEngine.Result, featureUnits: string[]): string[] {
        const metaExports = meta.exports;
        const metaDepGraph = meta.chunkDepGraph;
        const result: string[] = [];
        const visited = new Set<string>();
        const addChunk = (chunkFileName: string): void => {
            if (visited.has(chunkFileName)) {
                return;
            }
            visited.add(chunkFileName);
            result.push(chunkFileName);
            if (metaDepGraph && chunkFileName in metaDepGraph) {
                for (const dependencyChunk of metaDepGraph[chunkFileName]) {
                    addChunk(dependencyChunk);
                }
            }
        };
        for (const featureUnit of featureUnits) {
            const chunkFileName = metaExports[featureUnit];
            if (!chunkFileName) {
                console.error(`Feature unit ${featureUnit} is not in build result!`);
                continue;
            }
            addChunk(chunkFileName);
        }
        return result;
    }

    // eslint-disable-next-line no-inner-declarations
    function _enumerateDependentAssets (meta: buildEngine.Result, dependentChunks: string[]): string[] {
        const metaDepAsset = meta.assetDepGraph;
        let result: string[] = [];
        for (const chunkName of dependentChunks) {
            const depAssets = metaDepAsset[chunkName];
            if (depAssets?.length > 0) {
                result = result.concat(depAssets);
            }
        }
        return result;
    }

    /**
     * Enumerates all chunk files that used by specified feature units.
     * @param meta Metadata of build result.
     * @param featureUnits Feature units.
     */
    export function enumerateDependentChunks (meta: buildEngine.Result, featureUnits: string[]): string[] {
        return _enumerateDependentChunks(meta, featureUnits);
    }

    /**
     * Enumerates all asset files that used by specified feature units.
     * @param meta Metadata of build result.
     * @param featureUnits Feature units.
     */
    export function enumerateDependentAssets (meta: buildEngine.Result, featureUnits: string[]): string[] {
        const dependentChunks = _enumerateDependentChunks(meta, featureUnits);
        return _enumerateDependentAssets(meta, dependentChunks);
    }

    /**
     * Enumerates all chunk files and asset files that used by specified feature units.
     * @param meta Metadata of build result.
     * @param featureUnits Feature units.
     */
    export function enumerateAllDependents (meta: buildEngine.Result, featureUnits: string[]): string[] {
        const dependentChunks = _enumerateDependentChunks(meta, featureUnits);
        const dependentAssets = _enumerateDependentAssets(meta, dependentChunks);
        return dependentAssets.concat(dependentChunks);
    }
}