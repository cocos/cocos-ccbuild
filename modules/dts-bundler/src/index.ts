import fs from 'fs-extra';
import * as gift from '@cocos/tfig';
import { StatsQuery } from '@ccbuild/stats-query';
import { ps } from '@ccbuild/utils';
import { typescript as Transformer } from '@ccbuild/transformer';
import { ModuleQuery } from '@ccbuild/modularize';


import ts = Transformer.core;

const DEBUG = false;
const REMOVE_OLD = !DEBUG;
const RECOMPILE = !DEBUG;
const REMOVE_UNBUNDLED_CACHE = !DEBUG;

export interface Options {
    engine: string;
    outDir: string;
}

export async function build (options: Options): Promise<boolean> {
    console.log(`Typescript version: ${ts.version}`);

    const {
        engine,
        outDir,
    } = options;
    await fs.ensureDir(outDir);

    // TODO: should this be a build options ?
    const withIndex = true;
    const withExports = false;
    const withEditorExports = true;

    console.debug(`With index: ${withIndex}`);
    console.debug(`With exports: ${withExports}`);
    console.debug(`With editor exports: ${withEditorExports}`);

    const statsQuery = await StatsQuery.create(engine);
    const moduleQuery = new ModuleQuery({
        engine,
        platform: 'WEB_EDITOR',  // what ever platform is OK
        customExportConditions: ['types'],
    });
    const moduleExportMap = await moduleQuery.getExportMap();
    // NOTE: to record the imported modules, we only bundle the imported modules, not all engine modules.
    const importedModules: string[] = [];

    const tsConfigPath = statsQuery.tsConfigPath;

    const unbundledOutDir = ps.join(engine, '__dts_before_bundle');
    const parsedCommandLine = ts.getParsedCommandLineOfConfigFile(
        tsConfigPath, {
            declaration: true,
            noEmit: false,
            emitDeclarationOnly: true,
            outFile: undefined,
            outDir: unbundledOutDir,
        }, {
            onUnRecoverableConfigFileDiagnostic: () => {},
            useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
            readDirectory: ts.sys.readDirectory,
            getCurrentDirectory: ts.sys.getCurrentDirectory,
            fileExists: ts.sys.fileExists,
            readFile: ts.sys.readFile,
        },
    );
    if (!parsedCommandLine) {
        throw new Error(`Can not get 'parsedCommandLine'.`);
    }

    const unbundledOutDirNormalized = ps.resolve(engine, parsedCommandLine.options.outDir!);
    console.debug(`Unbundled will write to: ${unbundledOutDirNormalized}`);

    await fs.ensureDir(unbundledOutDirNormalized);
    if (REMOVE_OLD) {
        await fs.emptyDir(unbundledOutDirNormalized);
    }

    console.log(`Generating...`);

    const featureUnits = statsQuery.getFeatureUnits().filter((m) => m !== 'wait-for-ammo-instantiation');

    const editorExportModules = statsQuery.getEditorPublicModules();

    if (RECOMPILE) {
        let fileNames = parsedCommandLine.fileNames;
        if (withEditorExports) {
            fileNames = fileNames.concat(editorExportModules.map((e) => statsQuery.getEditorPublicModuleFile(e)));
        }

        const host = ts.createCompilerHost(parsedCommandLine.options);
        host.resolveModuleNames = (moduleNames, importer): (ts.ResolvedModule | undefined)[] => {
            const resolvedModules: (ts.ResolvedModule | undefined)[] = [];
            for (const moduleName of moduleNames) {
                const exportPath = moduleExportMap[moduleName];
                if (exportPath) {
                    if (!importedModules.includes(moduleName)) {
                        importedModules.push(moduleName);
                    }
                    resolvedModules.push({resolvedFileName: exportPath});
                } else {
                    const resolvedRes = ts.resolveModuleName(moduleName, importer, parsedCommandLine.options, host);
                    if (resolvedRes.resolvedModule) {
                        resolvedModules.push({resolvedFileName: resolvedRes.resolvedModule.resolvedFileName});
                    } else {
                        // Cannot resolve module, treat it as external.
                        resolvedModules.push(undefined);
                    }
                }
            }
            return resolvedModules;
        };
        const program = ts.createProgram(fileNames, parsedCommandLine.options, host);
        const emitResult = program.emit(
            undefined, // targetSourceFile
            undefined, // writeFile
            undefined, // cancellationToken,
            true, // emitOnlyDtsFiles
            undefined, // customTransformers
        );

        const allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
        for (const diagnostic of allDiagnostics) {
            let printer;
            switch (diagnostic.category) {
            case ts.DiagnosticCategory.Error:
                printer = console.error;
                break;
            case ts.DiagnosticCategory.Warning:
                printer = console.warn;
                break;
            case ts.DiagnosticCategory.Message:
            case ts.DiagnosticCategory.Suggestion:
            default:
                printer = console.log;
                break;
            }
            if (!printer) {
                continue;
            }
            if (diagnostic.file && diagnostic.start !== undefined) {
                const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
                const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, ts.sys.newLine);
                printer(`${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`);
            } else {
                printer(`${ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n')}`);
            }
        }
    }

    const patchSpineCoreDtsSource = ps.join(engine, 'cocos', 'spine', 'lib', 'spine-core.d.ts');
    const patchSpineCoreDtsTarget = ps.join(unbundledOutDirNormalized, 'cocos', 'spine', 'lib', 'spine-core.d.ts');
    if (!await fs.pathExists(patchSpineCoreDtsSource)) {
        console.debug(`Does 'cocos/spine/lib/spine-core.d.ts' no longer existed? I have a patch for it.`);
    } else {
        console.debug(`It's ${new Date().toLocaleString()}, we're still doing the hack for spine-core.d.ts`);
        await fs.ensureDir(ps.dirname(patchSpineCoreDtsTarget));
        await fs.copyFile(
            patchSpineCoreDtsSource,
            patchSpineCoreDtsTarget,
        );
    }

    const rebasedModuleExportMap: Record<string, string> = {};
    for (const [moduleName, modulePath] of Object.entries(moduleExportMap)) {
        let rebasedPath = ps.rebasePath(modulePath, engine, unbundledOutDirNormalized);
        if (!rebasedPath.endsWith('.d.ts')) {
            rebasedPath = ps.replaceExtname(rebasedPath, '.ts', '.d.ts');
        }
        rebasedModuleExportMap[moduleName] = rebasedPath;
    }

    const giftInputs: string[] = [];
    const types = parsedCommandLine.options.types?.map((typeFile) => `${typeFile}.d.ts`);
    if (types) {
        for (let file of types) {
            const isBareSpecifier = !file.includes('/');
            if (isBareSpecifier) {
                file = require.resolve(`@types/${file.slice(0, -'.d.ts'.length)}`, {
                    paths: [engine]
                });
            }
            if (!ps.isAbsolute(file)) {
                file = ps.join(ps.dirname(tsConfigPath), file);
            }
            giftInputs.push(file);
        }
    }
    const listGiftInputs = async (dir: string): Promise<void> => {
        for (const file of await fs.readdir(dir)) {
            const path = ps.join(dir, file);
            // eslint-disable-next-line no-await-in-loop
            const stats = await fs.stat(path);
            if (stats.isFile()) {
                giftInputs.push(path);
            } else if (stats.isDirectory()) {
                // eslint-disable-next-line no-await-in-loop
                await listGiftInputs(path);
            }
        }
    };
    await listGiftInputs(unbundledOutDirNormalized);

    const giftEntries: Record<string, string> = { };
    for (const [moduleName, modulePath] of Object.entries(rebasedModuleExportMap)) {
        if (importedModules.includes(moduleName)) {
            giftInputs.push(modulePath);
        }
    }

    const getModuleNameInTsOutFile = (moduleFile: string): string => {
        const path = ps.relative(statsQuery.path, moduleFile);
        const pathDts = path.replace(/\.ts$/, '.d.ts');
        return ps.join(unbundledOutDirNormalized, pathDts);
    };

    if (withExports) {
        for (const exportEntry of featureUnits) {
            giftEntries[exportEntry] = getModuleNameInTsOutFile(
                statsQuery.getFeatureUnitFile(exportEntry),
            );
        }
    }

    if (withEditorExports) {
        for (const editorExportModule of editorExportModules) {
            giftEntries[editorExportModule] = getModuleNameInTsOutFile(
                statsQuery.getEditorPublicModuleFile(editorExportModule),
            );
        }
        for (const [moduleName, modulePath] of Object.entries(rebasedModuleExportMap)) {
            if (moduleName.endsWith('/editor') && importedModules.includes(moduleName)) {
                const editorModuleName = transformToEditorModuleName(moduleName);
                giftEntries[editorModuleName] = modulePath;
            }
        }
    }

    let ccDtsFile: string | undefined;
    if (withIndex && !withExports) {
        ccDtsFile = ps.join(unbundledOutDirNormalized, 'virtual-cc.d.ts');
        giftEntries.cc = ccDtsFile;
        giftInputs.push(ccDtsFile);
        const code = `// Auto-generated\n${
            statsQuery.evaluateIndexModuleSource(featureUnits,
                (featureUnit) => getModuleNameInTsOutFile(statsQuery.getFeatureUnitFile(featureUnit)).replace(/\\/g, '/').replace(/\.d.ts$/, ''))
        }\n`;
        await fs.writeFile(ccDtsFile, code, { encoding: 'utf8' });
    }

    console.log(`Bundling...`);
    try {
        const indexOutputPath = ps.join(outDir, 'cc.d.ts');
        const giftResult = gift.bundle({
            input: giftInputs,
            rootDir: unbundledOutDirNormalized,
            name: 'cc',
            rootModule: 'index',
            entries: giftEntries,
            priority: [
                ...(ccDtsFile ? [ccDtsFile] : []), // Things should be exported to 'cc' as far as possible.
            ],
            privateJsDocTag: 'engineInternal',
            groups: [
                { test: /^cc\/editor.*$/, path: ps.join(outDir, 'cc.editor.d.ts') },
                { test: /^cc\/.*$/, path: ps.join(outDir, 'index.d.ts') },
                { test: /^cc.*$/, path: indexOutputPath },
            ],
            nonExportedSymbolDistribution: [{
                sourceModule: /cocos\/animation\/marionette/,
                targetModule: 'cc/editor/new-gen-anim',
            }, {
                sourceModule: /.*/, // Put everything non-exported that 'cc' encountered into 'cc'
                targetModule: 'cc',
            }],
            moduleMap: rebasedModuleExportMap,
        });

        await Promise.all(giftResult.groups.map(async (group) => {
            await fs.outputFile(group.path, group.code, { encoding: 'utf8' });
        }));

        if (withIndex && withExports) {
            await fs.outputFile(
                indexOutputPath,
                buildIndexModule(featureUnits, statsQuery),
                { encoding: 'utf8' },
            );
        }
    } catch (error) {
        console.error(error);
        return false;
    } finally {
        if (REMOVE_UNBUNDLED_CACHE) {
            await fs.remove(unbundledOutDirNormalized);
        }
    }

    return true;
}

function buildIndexModule (featureUnits: string[], statsQuery: StatsQuery): string {
    return `declare module "cc" {\n${
        statsQuery.evaluateIndexModuleSource(featureUnits)
            .split('\n')
            .map((line) => `    ${line}`)
            .join('\n')
    }\n}`;
}

/**
 * '@cocos/moduleName' -> 'cc/editor/moduleName'
 * '@cocos/moduleName/editor' -> 'cc/editor/moduleName'
 * 'moduleName' -> 'cc/editor/moduleName'
 * @param moduleName 
 */
function transformToEditorModuleName (moduleName: string): string {
    const split = moduleName.split('/');
    moduleName = split.length > 1 ? split[1] : split[0];
    return `cc/editor/${moduleName}`;
}
