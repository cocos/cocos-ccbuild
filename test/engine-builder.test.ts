import * as ccbuild from '../src/index'
import * as ps from 'path';

test('test base', async () => {
    const engineBuilder = new ccbuild.EngineBuilder();
    const root = ps.join(__dirname, './test-source').replace(/\\/g, '/');
    const buildResult = await engineBuilder.build({
        root,
        entries: [
          ps.join(__dirname, './test-source/exports/audio.ts').replace(/\\/g, '/'),
          ps.join(__dirname, './test-source/exports/animation.ts').replace(/\\/g, '/'),
        ],
        platform: 'OPEN_HARMONY',
        flagConfig: {
            DEBUG: true,
        },
        outDir: ps.join(__dirname, './lib').replace(/\\/g, '/'),
        virtualModule: {
          'virtual': 'export const TEST = true;\nexport const EDITOR = false;'
        },
    });
    const files = Object.keys(buildResult);
    for (let file of files) {
      const data = buildResult[file];
      delete buildResult[file];
      const relativeFile = ps.relative(root, file).replace(/\\/g, '/');
      buildResult[relativeFile] = data;
    }
    expect(buildResult).toMatchInlineSnapshot(`
{
  "__virtual__/virtual.ts": {
    "code": "export const TEST = true;
export const EDITOR = false;",
  },
  "animation/animation.ts": {
    "code": "export class Animation {
    play () {}
}",
  },
  "animation/index.ts": {
    "code": "export * from './animation'",
  },
  "audio/index.ts": {
    "code": "export * from './player';",
  },
  "audio/player.ts": {
    "code": "import { EDITOR, TEST } from "virtual";

export class Player {
    play () {
        if (EDITOR) {
            console.log('this is editor');
        } else if (TEST) {
            console.log('this is test');
        }
    }
}",
  },
  "exports/animation.ts": {
    "code": "export * from '../animation'",
  },
  "exports/audio.ts": {
    "code": "export * from '../audio'",
  },
}
`);
});