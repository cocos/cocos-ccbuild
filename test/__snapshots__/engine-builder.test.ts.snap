// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test base 1`] = `
{
  "__virtual__/internal_constants.ts": {
    "code": "function tryDefineGlobal(name, value) {
  const _global = typeof window === 'undefined' ? global : window;
  if (typeof _global[name] === 'undefined') {
    return _global[name] = value;
  } else {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return _global[name];
  }
}
export const HTML5 = false;
export const NATIVE = true;
export const WECHAT = false;
tryDefineGlobal('CC_WECHAT', false);
export const BAIDU = false;
tryDefineGlobal('CC_BAIDU', false);
export const XIAOMI = false;
tryDefineGlobal('CC_XIAOMI', false);
export const ALIPAY = false;
tryDefineGlobal('CC_ALIPAY', false);
export const TAOBAO = false;
export const BYTEDANCE = false;
tryDefineGlobal('CC_BYTEDANCE', false);
export const OPPO = false;
tryDefineGlobal('CC_OPPO', false);
export const VIVO = false;
tryDefineGlobal('CC_VIVO', false);
export const HUAWEI = false;
tryDefineGlobal('CC_HUAWEI', false);
export const COCOSPLAY = false;
tryDefineGlobal('CC_COCOSPLAY', false);
export const QTT = false;
tryDefineGlobal('CC_QTT', false);
export const LINKSURE = false;
tryDefineGlobal('CC_LINKSURE', false);
export const EDITOR = false;
tryDefineGlobal('CC_EDITOR', false);
export const PREVIEW = false;
tryDefineGlobal('CC_PREVIEW', false);
export const BUILD = true;
tryDefineGlobal('CC_BUILD', true);
export const TEST = false;
tryDefineGlobal('CC_TEST', false);
export const DEBUG = true;
tryDefineGlobal('CC_DEBUG', true);
export const SERVER_MODE = false;
export const DEV = false;
tryDefineGlobal('CC_DEV', false);
export const MINIGAME = false;
tryDefineGlobal('CC_MINIGAME', false);
export const RUNTIME_BASED = false;
tryDefineGlobal('CC_RUNTIME_BASED', false);
export const SUPPORT_JIT = true;
tryDefineGlobal('CC_SUPPORT_JIT', true);
export const JSB = true;
tryDefineGlobal('CC_JSB', true);
export const NOT_PACK_PHYSX_LIBS = false;
export const NET_MODE = 0;
export const WEBGPU = false;",
  },
  "animation/animation.ts": {
    "code": "export class Animation {
  play() {}
}",
  },
  "animation/index.ts": {
    "code": "export * from './animation';",
  },
  "audio/index.ts": {
    "code": "export * from './player';",
  },
  "audio/internal-native.ts": {
    "code": "console.log('internal-native');",
  },
  "audio/module.native.ts": {
    "code": "console.log('module that after override');",
  },
  "audio/player.ts": {
    "code": "import "internal-native.ts";
import { EDITOR, TEST } from "../__virtual__/internal_constants.ts";
import "./module.native";
import "../pal/audio/native.ts";
import "../pal/minigame/native.ts";
import { testDecorator } from "../decorators";
@testDecorator
export class Player {
  @testDecorator
  play() {
    if (EDITOR) {
      console.log('this is editor');
    } else if (TEST) {
      console.log('this is test');
    }
  }
}
export const player: Player = new Player();",
  },
  "decorators.ts": {
    "code": "export function testDecorator() {}",
  },
  "exports/animation.ts": {
    "code": "export * from '../animation';",
  },
  "exports/audio.ts": {
    "code": "export * from '../audio';",
  },
  "pal/audio/native.ts": {
    "code": "console.log('native audio');",
  },
  "pal/minigame/native.ts": {
    "code": "console.log('native');",
  },
}
`;
